[clang][diagnostics] Stable IDs for Clang diagnostics

dbartol opened the pull request:
Part of the implementation of [[RFC] Emitting Auditable SARIF Logs from Clang](https://discourse.llvm.org/t/rfc-emitting-auditable-sarif-logs-from-clang/88624)

SARIF diagnostics require that each rule have a stable `id` property to identify that rule across runs, even when the compiler or analysis tool has changed. We were previously setting the `id` property to the numeric value of the enum value for that diagnostic within the Clang implementation; this value changes whenever an unrelated diagnostic is inserted or removed earlier in the list.

This change sets the `id` property to the _text_ of that same enum value. This value would only change if someone renames the enum value for that diagnostic, which should happen much less frequently than renumbering.

For now, we will just assume that renaming happens infrequently enough that existing consumers of SARIF will not notice. In the future, we could take advantage of SARIF's support for `deprecatedIds`, which let a rule specify the IDs by which it was previously known. This would let us rename, split, or combine diagnostics while still being able to correlate the new diagnostic IDs with older SARIF logs and/or suppressions.

Nothing in this change affects how warnings are configured on the command line or in `#pragma clang diagnostic`. Those still use warning groups, not the stable IDs.

### Potential discussion topics
From @AaronBallman on the RFC:
>We believe some open questions remain (things like whether a unique ID is on the per-diagnostic level or on the diagnostic group level, whether the ID is explicitly spelled in the .td file or implicitly generated, whether we document the IDs, etc), but we think those questions are best decided in PR discussions with interested parties rather than an RFC.

As a starting point, this PR proposes the following answers to those open questions:
- _whether a unique ID is on the per-diagnostic level or on the diagnostic group level_ - per-diagnostic level. For my justification, see [this portion of the RFC discussion](https://discourse.llvm.org/t/rfc-emitting-auditable-sarif-logs-from-clang/88624/11?u=dbartol.).
- _whether the ID is explicitly spelled in the .td file or implicitly generated_ - Implicitly generated, but I'd be happy to have a way to explicitly specify it. I just think that the in-code identifier is a reasonable default, and manually reviewing the IDs of thousands of existing diagnostics would add little benefit.
- _whether we document the IDs_ - For now, the IDs are only exposed to the user (and other tools) in the SARIF file, so I don't think we need to document these. We could certainly add this information to the output of `diagtool` in the future if users find it relevant.

dbartol committed:
[clang] [diagnostics] Stable IDs for Clang diagnostics

SARIF diagnostics require that each rule have a stable `id` property to identify that rule across runs, even when the compiler or analysis tool has changed. We were previously setting the `id` property to the numeric value of the enum value for that diagnostic within the Clang implementation; this value changes whenever an unrelated diagnostic is inserted or removed earlier in the list.

This change sets the `id` property to the _text_ of that same enum value. This value would only change if someone renames the enum value for that diagnostic, which should happen much less frequently than renumbering.

For now, we will just assume that renaming happens infrequently enough that existing consumers of SARIF will not notice. In the future, we could take advantage of SARIF's support for `deprecatedIds`, which let a rule specify the IDs by which it was previously known. This would let us rename, split, or combine diagnostics while still being able to correlate the new diagnostic IDs with older SARIF logs and/or suppressions.

Nothing in this change affects how warnings are configured on the command line or in `#pragma clang diagnostic`. Those still use warning groups, not the stable IDs.

Sirraide reviewed (commented):
The implementation of this seems fine to me, but the RFC has to be accepted first before this can be merged.

I don’t think there is a good way to deal w/ custom diag ids. Those *shouldn’t* be used anywhere in Clang anyway (I know there are a few places that use them, but those should all be updated to use proper diagnostics instead).

dbartol commented:
@AaronBallman 
>should we auto-generate a UUID...

To persist across builds, the IDs would either have to be in the `.td` files, or they'd need to be generated by some hash of stable inputs. I don't think we have any stable inputs to hash - the message, internal ID, and integer value could all change. 

The SARIF stable ID doesn't have to be meaningful. If we went with UUIDs, someone adding a new diagnostic could just generate one easily enough. We'd just need a check somewhere to prevent duplicates in case of copy-paste mistakes.

We could also just use explicitly-assigned, dense-ish stable integer IDs, MSVC-style. Those are easier to use in both written and oral conversation than UUIDs, but they might require slightly more process. We'd do an initial pass to add them to the `.td` file, just like for UUIDs, and add the same check for duplicates. Adding a new diagnostic would require picking an unused integer ID. We'd probably have to keep a list of "retired" IDs in the code, though, to prevent accidental reuse of a no-longer-generated diagnostic's ID for a new diagnostic. The main problem with this approach would be dealing with forks and merge conflicts. If two forks both add warning 1234 and then try to push upstream, we'd have to have a rule like "first one into `llvm-project/main` wins, everyone else has to renumber".

I suppose we could split the difference and use one of those code phrase generators to make diagnostic IDs. Easier to use in conversation than UUIDs, but less likely to collide than manually-assigned dense integers. The downside is that at some point, we'll find ourselves writing something like "We're getting too many false positive reports for purple-monkey-dishwasher, so we'll need to tune the heuristcs."

Does any one of the above approaches seem workable?

dbartol commented:
@AaronBallman Your idea of having a tool to add a diagnostic got me thinking. Please bear with me...

I'll start with a couple of claims:

### Warnings only
We should only really care about stable IDs for warnings, not errors/fatals/notes. While SARIF is capable of representing errors, I don't expect anyone to try to track instances of errors over time. Either your code compiles successfully or it doesn't. If it has errors, you have to fix them. The only reason why I think we should emit errors to the SARIF log at all is that having both errors and warnings in the same file makes it easier to use SARIF as a common structured diagnostic format to drive IDE support, CI reports, etc. So, while SARIF requires us to emit a `ruleId` property for each result, whether warning or error, I don't think we should make any effort to make the _error_ IDs stable for now. 

We seem to use notes only to annotate other warnings/errors. In SARIF, those are supposed to show up as `relatedLocations` on the root warning/error, rather than as their own `result` object. Thus, notes don't need SARIF IDs at all, stable or otherwise.

### Focus on stability and evolution
The SARIF ID doesn't need to meaningful; it just needs to be unique and stable. So, instead of focusing on how to choose a stable ID for a warning, we should focus on how to ensure that the ID _remains_ stable.  We should make creation of a stable ID trivial, but then detect when new warnings are added or removed, and ensure that the developer has taken any necessary steps to preserve the stable IDs or to record how the stable IDs have evolved.

## Proposal
For errors, fatals, and notes, we use the in-source identifier of the diagnostic as its SARIF ID. We do not expect this to be stable, but we do not expect SARIF tools to care about the stability of error IDs over time.

For warnings, we use the in-source identifier of the diagnostic as its SARIF ID by default. We also allow two additional properties: `StableId<id>` and `OldStableIds<[id, ...]`. `StableId<id>` specifies _id_ as the stable ID of the diagnostic, instead of the default. `OldStableIds<[id, ...]>` adds the listed IDs to the `deprecatedIds` property of the diagnostic's SARIF rule metadata.

The effect on common contributor scenarios is as follows:

- **Adding a new diagnostic (either warning or non-warning):** Same as today. Just add it, and we'll use the in-source identifier as its stable ID.
- **Deleting an existing diagnostic (either warning or non-warning):** Same as today. Just delete it.
- **Modifying a property of an existing warning _other than its in-source identifier_:** Same as today.
- **Changing the in-source identifier of a warning:** Change the in-source identifier, but need to either explicitly specify `StableId<original-name>` (to keep the original stable ID) or `OldStableIds<[original-name]>` (to use the new in-source identifier as the stable ID while preserving history).
- **Splitting a warning into multiple warnings:** Create the new warnings as today, but add `OldStableIds<[original-name]>` to each of them to preserve history.
- **Combining multiple warnings into a single warning:** As today, but add `OldStableIds<[original-name-1, original-name-2...]>` to the combined warning to preserve history.

## Enforcement
It should be easy enough to create a CI job that compares the set of diagnostics between the base commit of a PR and the head commit of that PR, and, if anything suspicious changes have occurred, posts instructions for how to deal with renaming/merging/splitting warnings.

To give you an idea of how often this would trigger, I wrote a simple script to crawl our Git history, run `llvm-tblgen Diagnostic.td -dump-json` on each commit and its parent, and identify whether that commit added a new warning, deleted an existing warning, both, or neither. Crawling back from current `main` to the beginning of 2024, there were:

- 96 commits that added one or more new warnings without deleting any existing warnings. These could be splitting an existing warning, but are usually just new warnings.
- 15 commits that deleted one or more existing warnings without adding any new warnings. These could be merging existing warnings into one, but are usually just deleting warnings.
- 13 commits that both added and deleted warnings. These often appear to be doing something complex involving renaming, merging, or splitting.
- 240 commits that modified one of the `Diagnostic*.td` files without adding or deleting any warnings.

(details [here](https://gist.github.com/dbartol/e0ebbeed741bf2ef46cad6dcb7477072))

So, the bot would fire ~5 times a month, and actually provoke additional work from the contributor maybe once every month or two. Other than that, the whole diagnostic process would remain the same as today, which still seems like less work than managing GUIDs for every diagnostic.

dbartol committed:
[clang] [diagnostics] Use `normalize_sarif` for Clang SARIF testing

This change modifies the one existing Clang test for SARIF output to use the same `normalize_sarif` sed script that we use for diffing SARIF in the Clang Static Analyzer tests, rather than using a complicated set of `CHECK` statements. This did require changing the SARIF output to include line breaks.

dbartol commented:
Pushed an additional commit to reuse the Static Analyzer's `normalize_sarif` script for testing SARIF output.

dbartol committed:
Fix formatting

dbartol commented:
@AaronBallman I went ahead and implemented the support for explicitly specifying the stable ID, and for listing previous stable IDs, in a [separate PR](https://github.com/dbartol/llvm-project/pull/2) against this PR. Any thoughts about this approach to managing stable IDs over time?

AaronBallman commented:
> @AaronBallman Your idea of having a tool to add a diagnostic got me thinking. Please bear with me...
> 
> I'll start with a couple of claims:
> ### Warnings only
> 
> We should only really care about stable IDs for warnings, not errors/fatals/notes. While SARIF is capable of representing errors, I don't expect anyone to try to track instances of errors over time. 

I'm not certain I agree with this; there are reasons to track instances of errors over time (for example, gathering statistics about how often users run into a particular error, automatic reverts for certain kinds of errors, etc).

> Either your code compiles successfully or it doesn't. If it has errors, you have to fix them. The only reason why I think we should emit errors to the SARIF log at all is that having both errors and warnings in the same file makes it easier to use SARIF as a common structured diagnostic format to drive IDE support, CI reports, etc. So, while SARIF requires us to emit a `ruleId` property for each result, whether warning or error, I don't think we should make any effort to make the _error_ IDs stable for now.

I think we should make the error IDs stable from the start, along with warnings. Notes and remarks, however, don't seem to really have any kind of ID that I think would make sense to track. At least, I can't think of anything!

> We seem to use notes only to annotate other warnings/errors. In SARIF, those are supposed to show up as `relatedLocations` on the root warning/error, rather than as their own `result` object. Thus, notes don't need SARIF IDs at all, stable or otherwise.

+1

> ### Focus on stability and evolution
> 
> The SARIF ID doesn't need to meaningful; it just needs to be unique and stable. So, instead of focusing on how to choose a stable ID for a warning, we should focus on how to ensure that the ID _remains_ stable. We should make creation of a stable ID trivial, but then detect when new warnings are added or removed, and ensure that the developer has taken any necessary steps to preserve the stable IDs or to record how the stable IDs have evolved.

+1

> ## Proposal
> 
> For errors, fatals, and notes, we use the in-source identifier of the diagnostic as its SARIF ID. We do not expect this to be stable, but we do not expect SARIF tools to care about the stability of error IDs over time.
> 
> For warnings, we use the in-source identifier of the diagnostic as its SARIF ID by default. We also allow two additional properties: `StableId<id>` and `OldStableIds<[id, ...]`. `StableId<id>` specifies _id_ as the stable ID of the diagnostic, instead of the default. `OldStableIds<[id, ...]>` adds the listed IDs to the `deprecatedIds` property of the diagnostic's SARIF rule metadata.

I think this should apply to warnings and errors both.

> The effect on common contributor scenarios is as follows:
> 
>     * **Adding a new diagnostic (either warning or non-warning):** Same as today. Just add it, and we'll use the in-source identifier as its stable ID.
> 
>     * **Deleting an existing diagnostic (either warning or non-warning):** Same as today. Just delete it.
> 
>     * **Modifying a property of an existing warning _other than its in-source identifier_:** Same as today.
> 
>     * **Changing the in-source identifier of a warning:** Change the in-source identifier, but need to either explicitly specify `StableId<original-name>` (to keep the original stable ID) or `OldStableIds<[original-name]>` (to use the new in-source identifier as the stable ID while preserving history).
> 
>     * **Splitting a warning into multiple warnings:** Create the new warnings as today, but add `OldStableIds<[original-name]>` to each of them to preserve history.
> 
>     * **Combining multiple warnings into a single warning:** As today, but add `OldStableIds<[original-name-1, original-name-2...]>` to the combined warning to preserve history.

I think this makes sense. One question I have is: are these stable IDs usually displayed to the user in tools consuming SARIF? If so, are we sure these default IDs are reasonable for those cases? Or do we have to worry about tools expecting stable IDs to be short (so our long identifiers might get cut off such that the user only sees `warn_foo_w` and not `warn_foo_was_bad_because_bar`?

> ## Enforcement
> 
> It should be easy enough to create a CI job that compares the set of diagnostics between the base commit of a PR and the head commit of that PR, and, if anything suspicious changes have occurred, posts instructions for how to deal with renaming/merging/splitting warnings.
> 
> To give you an idea of how often this would trigger, I wrote a simple script to crawl our Git history, run `llvm-tblgen Diagnostic.td -dump-json` on each commit and its parent, and identify whether that commit added a new warning, deleted an existing warning, both, or neither. Crawling back from current `main` to the beginning of 2024, there were:
> 
>     * 96 commits that added one or more new warnings without deleting any existing warnings. These could be splitting an existing warning, but are usually just new warnings.
> 
>     * 15 commits that deleted one or more existing warnings without adding any new warnings. These could be merging existing warnings into one, but are usually just deleting warnings.
> 
>     * 13 commits that both added and deleted warnings. These often appear to be doing something complex involving renaming, merging, or splitting.
> 
>     * 240 commits that modified one of the `Diagnostic*.td` files without adding or deleting any warnings.
> 
> 
> (details [here](https://gist.github.com/dbartol/e0ebbeed741bf2ef46cad6dcb7477072))
> 
> So, the bot would fire ~5 times a month, and actually provoke additional work from the contributor maybe once every month or two. Other than that, the whole diagnostic process would remain the same as today, which still seems like less work than managing GUIDs for every diagnostic.

Thank you, that's a fantastic analysis of the impacts! I think that's quite reasonable.

dbartol commented:
@erichkeane 
>> _should all errors/warnings have explicit stable IDs?_
>
>Yes, there IS. This removes an entire class of 'rename problems' (AND the most likely ones!) to the point that PR-time enforcement will catch orders-of-magnitude fewer things. OF your list above, 96 + 15 + 240 all seem like 'not a problem'. This would enforce that breaking the SARIF name is an actual, active, intentional thing, which would remove all of the 13. I would still want periodic checking for these, but it would remove our risk of breaking them by almost 100%.

OK, let's do that. A couple details:
First, what's the best way to ensure that a developer changing or adding a diagnostic understands that the stable ID needs to be stable? Using the `StableId<>` class from my tblgen PR makes it more obvious, rather than just adding a second unnamed parameter to the `Warning<>` and `Error<>` classes. Does that sound OK?

Second, can I add those explicit stable IDs in a follow-up PR? So the PR sequence would be something like:
1. The current PR (implicit stable IDs)
2. Add tblgen support for (optional) explicit stable IDs [PR here](https://github.com/dbartol/llvm-project/pull/2)
3. Add explicit stable IDs for all warnings and errors.
4. Make explicit stable IDs required for warnings and errors.

I'm kind of dreading when PRs 3 and 4 hit my (or anyone else's) downstream fork, but I think having them as separate PRs will make that easier. It would be possible to merge PR 3 (which will probably have some merge conflicts), then go add stable IDs to all the downstream-only diagnostics, then merge PR 4.

erichkeane commented:
> @erichkeane
> 
> > > _should all errors/warnings have explicit stable IDs?_
> > 
> > 
> > Yes, there IS. This removes an entire class of 'rename problems' (AND the most likely ones!) to the point that PR-time enforcement will catch orders-of-magnitude fewer things. OF your list above, 96 + 15 + 240 all seem like 'not a problem'. This would enforce that breaking the SARIF name is an actual, active, intentional thing, which would remove all of the 13. I would still want periodic checking for these, but it would remove our risk of breaking them by almost 100%.
> 
> OK, let's do that. A couple details: First, what's the best way to ensure that a developer changing or adding a diagnostic understands that the stable ID needs to be stable? Using the `StableId<>` class from my tblgen PR makes it more obvious, rather than just adding a second unnamed parameter to the `Warning<>` and `Error<>` classes. Does that sound OK?
> 
Yep, agreed.  I suspect changing `StableId` is something that any human programmer would think twice about, and hopefully any reviewer would notice.  I think an implicit value is likely to not do a good job here.

> Second, can I add those explicit stable IDs in a follow-up PR? So the PR sequence would be something like:
> 
>     1. The current PR (implicit stable IDs)
> 
>     2. Add tblgen support for (optional) explicit stable IDs [PR here](https://github.com/dbartol/llvm-project/pull/2)
> 
>     3. Add explicit stable IDs for all warnings and errors.
> 
>     4. Make explicit stable IDs required for warnings and errors.
> 

I think a better order: Add support for `StableId`, add explicit IDs, make explicit `StableId` required for warnings/errors.  I don't see any value to the 'implicit' version in the meantime.

> 
> I'm kind of dreading when PRs 3 and 4 hit my (or anyone else's) downstream fork, but I think having them as separate PRs will make that easier. It would be possible to merge PR 3 (which will probably have some merge conflicts), then go add stable IDs to all the downstream-only diagnostics, then merge PR 4.

Yes, downstreams are going to have some pain here :)  Perhaps share the script you use for the explicit stable-id values/adding them would be a friendly thing to do, but that is the risk of being a downstream :)

dbartol commented:
@erichkeane 

> I think a better order: Add support for StableId, add explicit IDs, make explicit StableId required for warnings/errors. I don't see any value to the 'implicit' version in the meantime.

It's a bit more difficult to test the SARIF output without stable IDs. Without stable IDs, we were still using the integer value of the enum as the ID, which is obviously not stable, and had extra filtering in our tests to compensate. I don't think having the implicit stable IDs hurts anything in the short time it will take for me to write the update script and open the explicit stable IDs PR. Nobody will actually care whether the IDs remain stable during that time, except for the handful of diagnostics actually tests via SARIF output.

dbartol committed:
[clang] [diagnostics] `StableId` and `LegacyStableIds` properties for diagnostics

This change adds tblgen support for two new properties on `Diagnostic` objects:

- `StableId<id>` - Explicitly specifies the stable ID for a diagnostic. If not specified, the stable ID defaults to the name of the diagnostic's enum member as before. The stable ID is what gets emitted as the `id` property of the diagnostic's `Rule` object in SARIF.
- `LegacyStableIds<[id1, id2, ...]> - Specifies a list of stable IDs by which this diagnostic was previously known. These are emitted in the `deprecatedIds` property of the diagnostic's `Rule` object in SARIF. This property is used when the diagnostic's stable ID is renamed, when a diagnostic is split into multiple diagnostics, or when multiple diagnostics are merged into a single diagnostic.

Most of the code changes are in tblgen, in the code that loads the new properties from their static data tables, and in the SARIF emitter code. There were some further mechanical changes wherever the `DIAG` macro was invoked, since it now has two new parameters.

To test these changes, I found a warning that was recently renamed, and added an `LegacyStableIds` list with the previous name, which is what we would have to do for any renamed warning in the future. I then added a test that ensures that when this warning is emitted, the SARIF has the right `id` and `deprecatedIds` properties.

dbartol committed:
Merge branch 'main' into sarif/stable-ids

steakhal reviewed (commented):
Hey, I made it! I left a couple of tiny comments inline, but overall the PR looks very solid and well thought through! I was pleased to review this.

dbartol commented on clang/test/Frontend/Inputs/expected-sarif/sarif-legacy-stable-ids.c.sarif:
```diff
@@ -0,0 +1,76 @@
+  "version": "[SARIF version]"
+}
+
+1 warning generated.
```
The existing SARIF emission code (for `-fdiagnostics-format=sarif`) only supports emitting to stdout, so we're actually capturing additional console output after the SARIF is emitted. My next SARIF PR adds the ability to write the SARIF to a file, where we wouldn't see this extra output.

dbartol commented on clang/test/lit.cfg.py:
```diff
@@ -89,7 +89,26 @@
+        f"{sed_cmd} -r '%s;%s;%s;%s'"
+        % (
+            # Replace version strings that are likely to change.
+            r's/"version": "2.1.0"/"version": "[SARIF version]"/',
```
It is still relevant for future-proofing. This way when we move to SARIF 2.2.0 (or whatever), we can just update this pattern here, rather than all of our SARIF expectation files.

dbartol committed:
Fix PR feedback

dbartol commented on clang/test/Frontend/sarif-legacy-stable-ids.c:
```diff
@@ -0,0 +1,8 @@
+// RUN: %clang_cc1 -triple avr-unknown-unknown -fsyntax-only -fdiagnostics-format sarif %s > %t 2>&1 || true
```
I think that was a cut-n-paste from another test. I've removed the unnecessary `|| true`.

dbartol committed:
Expand comment

steakhal reviewed (approved):
This looks good to me. Thanks for addressing my feedback.
@AaronBallman how to you feel about this?
(I am also affiliated by the same company as Dave; I think we usually wait for external approvals around here)

dbartol committed:
Merge remote-tracking branch 'origin/main' into sarif/stable-ids

AaronBallman commented:
The only concern I had was regarding the behavior with custom diagnostics; so long as we don't trip the assert on every custom diagnostic, I think this is in good shape.
